prep/go/tutorial/chapter4/xkcd.go

```diff
@@ -0,0 +1,465 @@
+    if err != nil {
+        log.Fatal(err)
+        os.Exit(1)
+    }
+
+    for _, comic := range *comics {
+        if !comic.Indexed {
+            // index title and transcript
+            textToIndex := strings.Join([]string{comic.SafeTitle, comic.Title, comic.Transcript}, " ")
+            scanner := bufio.NewScanner(strings.NewReader(textToIndex))
+            scanner.Split(scanLetters)
+            for scanner.Scan() {
+                term := strings.ToLower(scanner.Text())
+                // Skip words like "a" and "to"
+                if len(term) < 3 {
+                    continue
+                }
+                addComicIndex(index, term, comic.Id)
+            }
+
+            // TODO flush comics back to disk with Index = True so it
+            // doesn't get indexed every run
+            // Trying to do it as everything is written now introduces a bug
+            // that isn't worth fixing until the performance is desired (profiling exercise? :))
+            comic.Indexed = true 
+        }
+    }
+    flushIndexToDisk(&index)
+    return index
+}
+
+// TODO ChatGPT'd this. Need to understand writing custom scanners outside this exercise
+func scanLetters(data []byte, atEOF bool) (advance int, token []byte, err error) {
+    start := 0
+    for ; start < len(data); start++ {
+        if unicode.IsLetter(rune(data[start])) {
+            break
+        }
+    }
+
+    for i := start; i < len(data); i++ {
+        if !unicode.IsLetter(rune(data[i])) {
+            return i + 1, data[start:i], nil
+        }
+    }
+
+    
+    if !atEOF && unicode.IsLetter(rune(data[len(data)-1])) {
+        return start, nil, nil
+    }
+
+    return len(data) + 1, data[start:], nil
+}
+
+func flushIndexToDisk(index *ComicIndex) {
+    data, err := json.MarshalIndent(index, "", " ")
+    if err != nil {
+        log.Fatal(err)
+    }
+
+    err = ioutil.WriteFile(COMIC_INDEX_STORAGE, data, 0644)
+    if err != nil {
+        log.Fatal(err)
+    }
+}
+
+
+func loadComics(comics *[]XKCDComic) {
+    var wg sync.WaitGroup
+    // Comics start at 1
+    i := 1
+    finished := false
+    for {
+        // Very funny xkcd.. Very funny..
+        if i == 404 {
+            i++
+        }
+
+        wg.Add(1)
+        // id is incrementing and passed by value so it is safe
+        // each routine will only access its index in the array
+        go tryLoadFromDiskFallbackToInternet(i, &(*comics)[i], &wg, &finished)
+
+        // Run lots of go routines for the first 2500
+        // Then slow down once we go past 2500
+        // If we let this go nuts we may start to timeout
+        if i < 2500 {
+            if i % 500 == 0 {
+                wg.Wait()
+            }
+        } else {
+            if i % 10 == 0 {
+                wg.Wait()
+            }
+        }
+
+        if finished {
+            // A routine hit the end of the comic list
+            // This is safe because even if multiple routines touch it
```

Comment:
This kind of thing can lead to subtle concurrency bugs; for example, see https://go.dev/ref/mem for way more information than you ever wanted to know.

You should always be using some sort of synchronization mechanism (channel, mutex, atomic variable) when coordinating between goroutines like this.

Reply:
I thought given I am only doing false -> true that this was safe.
That is because its different than say, incrementing a number, or having a boolean that can be flipped back and forth, locking isn't important.

This works also because I know in a given set of running goroutines. I will wait for them all to finish and if any marked the boolean as true I know I am done.

I think what I am most curious about is does this introduce a bug in this program and I can't reason about it or is it that this is something to not get in the habit of in general for the reasons you cited above.

Reply:
tl;dr if you do not use synchronization, crazy things can happen. I think it's best to show with an example.

The linked code snippet below will repeatedly initialize `x, y, r1, r2` to 0 and then run the following two goroutines concurrently:

Goroutine 1:

```
x = 1
r1 = y
```

Goroutine 2:

```
y = 1
r2 = x
```

My question to you (I strongly recommend thinking about this before opening and running the attachment): What are the possible final values of r1 and r2?

Link for code snippet:  https://github.com/bradfield-csi-5/johndavidmartinez/files/11932735/race.zip

Reply:
Both assignments happen first. Context switch between assignments.
x = 1
r1 = 1
y = 1
r2 = 1
Goroutine 1 then Goroutine 2
x = 1
r1 = 0
y = 1
r2 = 1
Goroutine 2 then Goroutine 1
x = 1
r1 = 1
y = 1
r2 = 0

So to answer the question possibilities:
r1 = 1;   r1 = 0;   r1 = 1;
r2 = 1;   r2 = 1;   r2 = 0;

I can't think of a situation where they're both 0.

Reply:
Ah. Ran the code. Interesting. So I can't really think of an assignment as a single operation here.

In the same way that incrementing a number isn't atomic because it is something like:
```
mov r1 eax
add 0x01 eax
mov eax r1
```

There's more happening in these assignments than I can reason about. I have no intuition for how this assignment race condition happens the 0, 0 case. It is a really good illustration on how a seemly simple assignment isn't threadsafe.

Reply:
One thing that can go wrong, for example, is that if the two goroutines run on two separate CPUs, then the updated values of `x` and `y` might only be in one of the CPU's caches, and won't be visible if the other CPU tries to read the value from main memory. Using a synchronization primitive such as a channel or mutex will eliminate these kinds of issues.